File: perf/benchmarks/apps/conway/conway.go
// conway — Game of Life N-step application benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

const W = 200

func main() {
	steps := 200
	if len(os.Args) > 1 {
		v, _ := strconv.Atoi(os.Args[1])
		steps = v
	}
	cells := W * W
	grid := make([]byte, cells)
	nxt := make([]byte, cells)
	var seed int64 = 1
	for i := 0; i < cells; i++ {
		seed = (seed*1103515245 + 12345) & 0x7FFFFFFF
		if (seed>>16)&3 < 2 {
			grid[i] = 1
		}
	}

	t0 := time.Now()
	for s := 0; s < steps; s++ {
		for y := 0; y < W; y++ {
			yu := (y + W - 1) % W
			yd := (y + 1) % W
			row := y * W
			rowU := yu * W
			rowD := yd * W
			for x := 0; x < W; x++ {
				xl := (x + W - 1) % W
				xr := (x + 1) % W
				n := grid[rowU+xl] + grid[rowU+x] + grid[rowU+xr] +
					grid[row+xl] + grid[row+xr] +
					grid[rowD+xl] + grid[rowD+x] + grid[rowD+xr]
				alive := grid[row+x]
				var nv byte
				if alive != 0 {
					if n == 2 || n == 3 {
						nv = 1
					}
				} else if n == 3 {
					nv = 1
				}
				nxt[row+x] = nv
			}
		}
		grid, nxt = nxt, grid
	}
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	var cs int64 = 0
	for _, c := range grid {
		cs += int64(c)
	}
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/apps/conway/conway.rs
// conway — Game of Life N-step application benchmark (Rust).
use std::env;
use std::time::Instant;

const W: usize = 200;

fn main() {
    let args: Vec<String> = env::args().collect();
    let steps: i64 = if args.len() > 1 { args[1].parse().unwrap_or(200) } else { 200 };

    let cells = W * W;
    let mut grid: Vec<u8> = vec![0; cells];
    let mut nxt: Vec<u8> = vec![0; cells];
    let mut seed: i64 = 1;
    for i in 0..cells {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7FFF_FFFF;
        grid[i] = if ((seed >> 16) & 3) < 2 { 1 } else { 0 };
    }

    let t0 = Instant::now();
    for _ in 0..steps {
        for y in 0..W {
            let yu = (y + W - 1) % W;
            let yd = (y + 1) % W;
            let row = y * W;
            let row_u = yu * W;
            let row_d = yd * W;
            for x in 0..W {
                let xl = (x + W - 1) % W;
                let xr = (x + 1) % W;
                let n = grid[row_u + xl] + grid[row_u + x] + grid[row_u + xr]
                      + grid[row + xl] + grid[row + xr]
                      + grid[row_d + xl] + grid[row_d + x] + grid[row_d + xr];
                let alive = grid[row + x];
                nxt[row + x] = if alive != 0 {
                    if n == 2 || n == 3 { 1 } else { 0 }
                } else if n == 3 { 1 } else { 0 };
            }
        }
        std::mem::swap(&mut grid, &mut nxt);
    }
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    let cs: i64 = grid.iter().map(|&c| c as i64).sum();
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
